js/DropZone/DropZone.js

```diff
@@ -0,0 +1,289 @@
+     * @param {Event} event
+     */
+    handleWindowEnter (event) {
+        const onWindow = this.fileOnWindow(event.clientX, event.clientY);
+        const onDropZone = this.node.contains(onWindow);
+        const files = event.dataTransfer.items || event.dataTransfer.files;
+
+        // clear idle timer
+        this.idleTimer.clear();
+
+        // if we are unable to get a length from the files we know data transfer
+        // isn't supported, so we'll set this flag to use later
+        if (!files.length) {
+            this.supportsDataTransfer = false;
+        }
+
+        if (onWindow && !this.windowActive) {
+            // handle files on window
+            const {valid, text} = this.validator.validate(files, this.files.length, this.size);
+
+            this.windowActive = true;
+            this.callbackManager.create(
+                this.options.windowEnter,
+                this,
+                {valid, text}
+            );
+        }
+
+        if (onDropZone && !this.dropZoneActive) {
+            // handle files on DropZone
+            this.handleDropZoneEnter(files);
+        }
+    }
+
+    /**
+     * Handle a file being dragged off of the window
+     * @param {Event} event
+     * @param {Boolean} force
+     */
+    handleWindowLeave (event, force = false) {
+        const onDropZone = this.node.contains(document.elementFromPoint(event.clientX, event.clientY));
+        const onWindow = this.fileOnWindow(event.clientX, event.clientY);
+        const files = event.dataTransfer.items || event.dataTransfer.files;
+
+        if (force || !onWindow) {
+            this.windowActive = false;
+            this.dropZoneActive = false;
+            this.idleTimer.clear();
+            this.callbackManager.create(this.options.windowLeave, this);
+        } else if (!onDropZone && this.dropZoneActive) {
+            this.handleDropZoneLeave(files);
+        }
+    }
+
+    /**
+     * Handle drag entering the DropZone
+     * @param {Array} files
+     */
+    handleDropZoneEnter (files) {
+        const { valid, text } = this.validator.validate(files, this.getFiles().length, this.getSize());
+
+        this.idleTimer.clear();
+        this.dropZoneActive = true;
+        this.callbackManager.create(
+            this.options.dropZoneEnter,
+            this,
+            { valid, text }
+        );
+    }
+
+    /**
+     * Handle drag leaving the DropZone
+     * @param {Array} files
+     */
+    handleDropZoneLeave (files) {
+        const { valid, text } = this.validator.validate(files, this.getFiles().length, this.getSize());
+
+        this.idleTimer.clear();
+        this.dropZoneActive = false;
+        this.callbackManager.create(
+            this.options.dropZoneLeave,
+            this,
+            { valid, text }
+        );
+    }
+
+    /**
+     * Return all files in the store or a file at a specified index
+     * @param {Number} index
+     * @return {Array|Object}
+     */
+    getFiles (index = -1) {
```

Comment:
Changed my mind on this, all of the `getFiles` methods are split down into `getFiles()` & `getFile(index)`